stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/listener/internal/ThreadQueryListenerFull.kt

```diff
@@ -0,0 +1,135 @@
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.offline.plugin.listener.internal
+
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.extensions.internal.users
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.persistance.repository.MessageRepository
+import io.getstream.chat.android.client.persistance.repository.UserRepository
+import io.getstream.chat.android.client.plugin.listeners.ThreadQueryListener
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.client.utils.onSuccessSuspend
+import io.getstream.chat.android.offline.plugin.logic.channel.thread.internal.ThreadLogic
+import io.getstream.chat.android.offline.plugin.logic.internal.LogicRegistry
+import io.getstream.logging.StreamLog
+
+/**
+ * ThreadQueryListenerFull handles both state and database. It uses, if available, the database
+ * to update, if available, the state.
+ *
+ * @param logic [LogicRegistry] Optional class to handle state updates
+ * @param messageRepository [MessageRepository] Optional to handle database updates related to messages
+ * @param userRepository [UserRepository]  Optional to handle database updates related to user
+ * @param chatClient [ChatClient]
+ */
+internal class ThreadQueryListenerFull(
+    private val logic: LogicRegistry?,
+    private val messageRepository: MessageRepository?,
+    private val userRepository: UserRepository?,
+    private val chatClient: ChatClient,
+) : ThreadQueryListener {
+
+    private val logger = StreamLog.getLogger("Chat:ThreadQueryListener")
+
+    override suspend fun onGetRepliesPrecondition(messageId: String, limit: Int): Result<Unit> {
+        val loadingMoreMessage = logic?.thread(messageId)?.isLoadingMessages()
+
+        return if (loadingMoreMessage == true) {
+            val errorMsg = "already loading messages for this thread, ignoring the load requests."
+            logger.i { errorMsg }
+            Result(ChatError(errorMsg))
+        } else {
+            Result.success(Unit)
+        }
+    }
+
+    override suspend fun onGetRepliesRequest(messageId: String, limit: Int) {
+        val threadLogic = logic?.thread(messageId)
+
+        threadLogic?.setLoading(true)
+        val messages = messageRepository?.selectMessagesForThread(messageId, limit)
```

Comment:
It feels like in the case of State of this Listener we should look for the message inside the `ChannelState` first.